cmd/api/handlers-queries.go

```diff
@@ -0,0 +1,174 @@
+package main
+
+import (
+	"encoding/json"
+	"log"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/jmpsec/osctrl/pkg/queries"
+	"github.com/jmpsec/osctrl/pkg/settings"
+	"github.com/jmpsec/osctrl/pkg/utils"
+)
+
+// GET Handler to return a single query in JSON
+func apiQueryShowHandler(w http.ResponseWriter, r *http.Request) {
+	incMetric(metricAPIReq)
+	utils.DebugHTTPDump(r, settingsmgr.DebugHTTP(settings.ServiceAPI), false)
+	vars := mux.Vars(r)
+	// Extract name
+	name, ok := vars["name"]
+	if !ok {
+		incMetric(metricAPIErr)
+		apiErrorResponse(w, "error getting name", nil)
+		return
+	}
+	// Get query by name
+	query, err := queriesmgr.Get(name)
+	if err != nil {
+		incMetric(metricAPIErr)
+		if err.Error() == "record not found" {
+			log.Printf("query not found: %s", name)
+			apiHTTPResponse(w, JSONApplicationUTF8, http.StatusNotFound, ApiErrorResponse{Error: "query not found"})
+		} else {
+			apiErrorResponse(w, "error getting query", err)
+		}
+		return
+	}
+	// Serialize and serve JSON
+	apiHTTPResponse(w, JSONApplicationUTF8, http.StatusOK, query)
+	incMetric(metricAPIOK)
+}
+
+// POST Handler to run a query
+func apiQueriesRunHandler(w http.ResponseWriter, r *http.Request) {
+	incMetric(metricAPIReq)
+	utils.DebugHTTPDump(r, settingsmgr.DebugHTTP(settings.ServiceAPI), false)
+	var q DistributedQueryRequest
+	// Parse request JSON body
+	if err := json.NewDecoder(r.Body).Decode(&q); err != nil {
+		incMetric(metricAPIErr)
+		apiErrorResponse(w, "error parsing POST body", err)
+		return
+	}
+	// FIXME check validity of query
```

Comment:
This is actually quite difficult and might be something you don't want to do because it'll force you to keep in lock step with every device interfacing with osctrl.

What if osquery upgrades SQLite and it adds a new keyword? Now no queries that use that can run through osctrl even though they would normally work. What about different versions of osquery within the same infra, different tables.

I would say not to even worry about this and just make osctrl an extremely reliable middle man :)

Reply:
👍